### PR TITLE
Add new  ‘enumWithDescriptions’ Custom Property to allow adding Enum Keys with Description

### PR DESCRIPTION
**Pull request by sonika-shah**
### Describe your changes:

Introducing a new customProperty `enumWithDescriptions ` , this will allow us to add additional metadata to our enum keys. Initially, we are adding a description field to each key, but this setup allows for future expansions.
With `enumWithDescriptions`  -
- We have the option of ‘multi-select’ to allow adding multiple or single keys.
- We can  have only unique keys , duplicate keys will not be allowed.
-  Once defined, `enumWithDescriptions` keys cannot be removed; however, their descriptions can be modified
- Sample `enumWithDescriptions ` customProperty-
```
{
  "name": "enumWithDescriptionsPropertyWithDiffKey",
  "propertyType": {
    "id": "ff839fdf-1675-4d89-b352-ebbc06f55f70",
    "type": "type"
  },
  "description": "custom Property description",
  "customPropertyConfig": {
    "config": {
      "multiSelect": true,
      "values": [
        {
          "key": "Enum With Descriptions Key1",
          "description": "Key1 description"
        },
        {
          "key": "Enum With Descriptions Key2",
          "description": "Key2 description"
        }
      ]
    }
  }
}
```

#
### Type of change:

- [ ] Bug fix
- [x] Improvement
- [x] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [ ] My PR title is `Fixes <issue-number>: <short explanation>`
- [x] I have commented on my code, particularly in hard-to-understand areas. 
- [x] For JSON Schema changes: I updated the migration scripts or explained why it is not needed.

### CODE REVIEW

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java [hunk 1]
```diff
@@ -1421,41 +1425,100 @@ private void validateExtension(T entity) {
...
+        }
+        // Replace each metaEnum key in the fieldValue with the corresponding object from the
+        // propertyConfig
+        Map<String, JsonNode> keyToObjectMap =
+            StreamSupport.stream(propertyConfigNode.get("values").spliterator(), false)
+                .collect(Collectors.toMap(node -> node.get("key").asText(), node -> node));
+
+        if (fieldValue.isArray()) {
+          ArrayNode newArray = JsonNodeFactory.instance.arrayNode();
+
```
Comment (Siddhanttimeline) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java:1497]: we can use `JsonUtils` functions here as well I believe.
`ArrayNode newArrayNode = JsonUtils.getObjectNode().arrayNode();`

Comment (sonika-shah) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java]: thanks @Siddhanttimeline for the suggestion, have added the change

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java [hunk 2]
```diff
@@ -1421,41 +1424,97 @@ private void validateExtension(T entity) {
...
+        formatter = DateTimeFormatter.ofPattern(propertyConfig);
+        LocalDateTime dateTime = LocalDateTime.parse(fieldValue.textValue(), formatter);
+        jsonNode.put(fieldName, dateTime.format(formatter));
+        break;
+      case "time-cp":
+        formatter = DateTimeFormatter.ofPattern(propertyConfig);
+        LocalTime time = LocalTime.parse(fieldValue.textValue(), formatter);
+        jsonNode.put(fieldName, time.format(formatter));
+        break;
+      case "metaEnum":
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java:1480]: lets call it EnumWithDescriptions

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java [hunk 3]
```diff
@@ -1421,41 +1424,97 @@ private void validateExtension(T entity) {
...
 
+  private void validateAndUpdateExtensionBasedOnPropertyType(
+      T entity,
+      ObjectNode jsonNode,
+      String fieldName,
+      JsonNode fieldValue,
+      String customPropertyType,
+      String propertyConfig) {
+    DateTimeFormatter formatter;
+    switch (customPropertyType) {
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java:1457]: use Java17 syntax here plase

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java (hunk 2 above)
Comment (harshach) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java:1480]: How are we handling exceptions, what happens if one of these fails parsing?

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/TypeRepository.java [hunk 4]
```diff
@@ -187,6 +190,27 @@ private void validateProperty(CustomProperty customProperty) {
           throw new IllegalArgumentException("Enum Custom Property Type must have EnumConfig.");
         }
       }
+      case "metaEnum" -> {
+        CustomPropertyConfig config = customProperty.getCustomPropertyConfig();
+        if (config != null) {
+          MetaEnumConfig metaEnumConfig =
+              JsonUtils.convertValue(config.getConfig(), MetaEnumConfig.class);
+          if (metaEnumConfig == null
```
Comment (harshach) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/TypeRepository.java:198]: can we create validate method and handle these checks there?

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java (hunk 3 above)
Comment (sonika-shah) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java:1457]: updated the Java17 syntax

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/TypeRepository.java (hunk 4 above)
Comment (sonika-shah) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/TypeRepository.java]: added new validate method

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java (hunk 2 above)
Comment (sonika-shah) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/EntityRepository.java]: have created methods to handle respective types and their exception -
private void validateAndUpdateExtensionBasedOnPropertyType(
    T entity, ObjectNode jsonNode, String fieldName, JsonNode fieldValue, 
    String customPropertyType, String propertyConfig) {

    switch (customPropertyType) {
        case "date-cp", "dateTime-cp", "time-cp" -> {
            String formattedValue = getFormattedDateTimeField(
                fieldValue.textValue(), customPropertyType, propertyConfig, fieldName);
            jsonNode.put(fieldName, formattedValue);
        }
        case "enumWithDescriptions" -> handleEnumWithDescriptions(
            fieldName, fieldValue, propertyConfig, jsonNode, entity);
        default -> {}
    }
}

**Review by harshach — DISMISSED** [review]

**Review by Sachin-chaurasiya — APPROVED** [review]

**Review by chirag-madlani — APPROVED** [review]

**Review by Sachin-chaurasiya — APPROVED** [review]

**Review by chirag-madlani — APPROVED** [review]

**Pull request merged**